enhancements/update/targeted-update-edge-blocking.md

```diff
@@ -0,0 +1,720 @@
+
+#### Stranding supported clusters
+
+As described [in the documentation][support-documentation], supported updates are still supported even if incoming edges are blocked, and Red Hat will eventually provide supported update paths from any supported release to the latest supported release in its z-stream.
+There is a risk, with the dynamic, per-cluster graph, that targeted edge blocking removes all outgoing update recommendations for some clusters on supported releases.
+This risk can be mitigated in at least two ways:
+
+* For the fraction of customer clusters that do not [opt-out of submitting Insights/Telemetry][uploaded-telemetry-opt-out], we can monitor [the existing `cluster_version_available_updates`][uploaded-telemetry-cluster_version_available_updates] to check for clusters running older versions which are still reporting no available, recommended updates.
+
+* We can process the graph with tooling that removes all `conditionalEdges` and look for any supported versions without exit paths.
+
+#### Malicious conditions
+
+An attacker who compromises a cluster's [`upstream`][api-upstream] update service can already do some fairly disruptive things, like recommend updates from 4.1.0 straight to 4.8.0.
+But at the moment, the cluster administrator (or whoever is writing to [ClusterVersion's `spec.desiredUpdate`][api-desiredUpdate]) is still in the loop deciding whether or not to accept the recommendation.
+
+With this enhancement, the cluster-version operator will begin performing more in-cluster actions automatically, such as evaluating PromQL recommended by the upstream update service.
+If the Prometheus implementation is not sufficiently hardened, malicious PromQL might expose the cluster to the attacker, with the simplest attacks being computationally intensive queries which cost CPU and memory resources that the administrator would rather be spending on more important tasks.
+[Future monitoring improvements][mon-1772] might reduce the risk of expensive queries.
+And it's also possible to teach the cluster-version operator to reject PromQL that does not match expected patterns.
+
+#### Clusters without local Prometheus stacks
+
+The Prometheus and monitoring stacks are fairly resource-intensive.
+There are [open proposals][mon-1569] to reduce their resource requirements.
+It is possible that some future clusters decide they need to drop the Prometheus stack entirely, which would leave the CVO unable to evaluate conditions based on PromQL.
+A future mitigation would be extending to support [non-PromQL filters](#non-promql-filters).
+
+## Design Details
+
+### Test Plan
+
+[The graph-data repository][graph-data] should grow a presubmit test to enforce as much of the new schema as is practical.
+Validating any PromQL beyond "it's a string" is probably more trouble than its worth, but we should keep an eye on Telemetry to see if any deployed cluster-version operators are reporting trouble with condition evaluation.
```

Comment:
It seems like we're much more likely to have some sort of expression typo than to ask for the wrong field entirely. Could we reuse the parser in Prometheus so we can be sure the strings at least parse to a valid PromQL expression?

Reply:
I've pushed 07db299 -> 6330580,  committing to getting graph-data's PromQL in front of the Go PromQL parser in a graph-data presubmit.  A new *Alternatives* subsection discusses the tradeoffs.  I'm not all that picky about how this gets implemented, maybe we can be clever and somehow avoid vendoring a bunch of Go in graph-data.  Maybe it's not worth the trouble of being clever.  Maybe it wouldn't actually require all that much vendored code.  I don't think the risk here is high enough to need to bottom all of this out before landing the enhancement.